asteroid/dsp/beamforming.py

```diff
@@ -248,6 +248,91 @@ def compute_beamforming_vector(target_scm: torch.Tensor, noise_scm: torch.Tensor
         return bf_vect
 
 
+class GEVDBeamformer(Beamformer):
+    """Generalized eigenvalue decomposition speech distortion weighted multichannel Wiener filter.
+
+        Compare to SDW-MWF, spatial covariance matrix are computed from low rank approximation
+        based on eigen values decomposition,
+        see equation 62 in `[1] <https://hal.inria.fr/hal-01390918/file/14-1.pdf>`_.
+
+    Attributes:
+        mu (float): Speech distortion constant.
+        rank (int): Rank for the approximation of target covariance matrix,
+            no approximation is made if rank is None.
+
+    References:
+        [1] R. Serizel, M. Moonen, B. Van Dijk and J. Wouters,
+        "Low-rank Approximation Based Multichannel Wiener Filter Algorithms for
+        Noise Reduction with Application in Cochlear Implants,"
+        in IEEE/ACM Transactions on Audio, Speech, and Language Processing, April 2014.
+    """
+
+    def __init__(self, mu: float, rank: int):
+        self.mu = mu
+        self.rank = rank
+
+    def compute_beamforming_vector(self, target_scm: torch.Tensor, noise_scm: torch.Tensor):
+        """Compute beamforming vectors for GEVD beamFormer.
+
+        Args:
+            target_scm (torch.ComplexTensor): shape (batch, mics, mics, freqs)
+            noise_scm (torch.ComplexTensor): shape (batch, mics, mics, freqs)
+
+        Returns:
+            torch.ComplexTensor: shape (batch, mics, freqs)
+
+        """
+        #  GEV decomposition of noise_scm^(-1) * target_scm
+        e_values, e_vectors = _generalized_eigenvalue_decomposition(
+            target_scm.permute(0, 3, 1, 2),  # bmmf -> bfmm
+            noise_scm.permute(0, 3, 1, 2),  # bmmf -> bfmm
+        )
+
+        #  Prevent negative and infinite eigenvalues
+        eps = torch.finfo(e_values.dtype).eps
+
+        e_values = torch.maximum(e_values, eps * torch.ones(e_values.shape))
+        e_values = torch.minimum(e_values, 1e6 * torch.ones(e_values.shape))
```

Comment:
Does this work on GPU? It looks like the ones would be created on a different device. 
Anyway, I think this is equivalent to `torch.clamp(e_values, min=eps, max=1e6)` which doesn't have device problems, right?

Reply:
You're right, this don't work on the GPU, because tensor `eps * torch.ones(e_values.shape)` has to be moved on the GPU,
and `torch.clamp()` is more suitable.